fetch.bs

```diff
@@ -2204,6 +2204,39 @@ unset or <a for=request>keepalive</a> is false, <a lt=terminated for=fetch>termi
 <a for="fetch group">fetch record</a>'s <a for="fetch record">fetch</a>.
 
 
+<h3 id=resolving-domains>Resolving domains</h3>
+
+<p>To <dfn>resolve a domain</dfn>, given a <a for=/>domain</a> <var>domain</var> and a
+<a for=/>network partition key</a> <var>key</var>, perform an <a>implementation-defined</a>
+operation to turn <var>domain</var> into a <a for=/>set</a> of one or more
+<a for=/>IP addresses</a>. If this operation succeeds, return the <a for=/>set</a> of
+<a for=/>IP addresses</a>. If it fails, return failure. The results of this operation may be cached.
+If they are cached, <var>key</var> must be used as part of the cache key.
+
+<p class=note>Typically this operation would involve DNS. The particulars (apart from the cache key)
+are not tied down as they are not pertinent to the system the Fetch Standard establishes. Other
+documents ought not to build on this primitive without having a considered discussion with the Fetch
+Standard community first. [[RFC1035]]
+
+<p>To <dfn>resolve an origin</dfn>, given an <a for=/>origin</a> <var>origin</var> and a
+<a for=/>network partition key</a> <var>key</var>:
+<!-- Should we assert the scheme here to be an HTTP(S) scheme or a WebRTC scheme? -->
+
+<ol>
+ <li><p>If <var>origin</var>'s <a for=origin>host</a> is an <a for=/>IP address</a>, then return
+ « <var>origin</var>'s <a for=origin>host</a> ».
+
+ <li><p>If the user agent is configured to use a proxy, then return
+ « <var>origin</var>'s <a for=origin>host</a> ».
```

Comment:
In the case of a proxy, do we still need to resolve IP addresses?  I don't believe Chrome currently does (in, e.g., the RFC 1918 case)

Reply:
Do you mean resolving the domain? This avoids doing that, right?

Reply:
No, I mean resolve the IP address.  e.g., mapping https://127.0.0.1/ to 0xFF000001 (sorry, may have the order reversed).  Chrome, at least, only tracks the proxy's IP in this case, and only provides that to consumers.  Consumers are careful not to treat it as the server IP.  So this would require a rearchitecture a number of Chrome layers.  Consumers would either have to extract the IP from the URL in the proxy case, or Chrome's network stack would have to provide both IPs, and all consumers audited.

Reply:
It depends on the type of proxy being used. We do resolve hosts for some proxies (e.g. SOCKSv4), but not for others (e.g. HTTP CONNECT). And we may still need to resolve prior to a proxy being involved (e.g. a PAC script)

I think we should strip the proxy-particular language here with regards to resolving; it's more closely tied to the connection establishment logic, and I think we'd have to tackle it there. That said, I'm not 100% sure on that - I think it really depends on how we end up using this "method" here that dictates which parts are captured in the pseudocode.

Reply:
Note that while we resolve server IPs for SOCKS4, we don't expose them to consumers.  My main concern here is the impact on any features that want the server IP from a socket (RFC1918 is the only one I'm aware of, but there may be others).

Reply:
Gotcha. Yeah, I believe you're talking about https://github.com/WICG/private-network-access/issues/52 . 

I don't believe Chromium's behaviour here were particularly opinionated, so much as it being an implementation detail, but tagging @letitz to review :) You're right, though, that the current logic prefers the IP address of the proxy server, at least when deciding Private Network Access resolution, even though it may use the IP address of the host to connect through a proxy.

Reply:
I think we have an "unknown" value in the proxy case, with its own special handling, rather than using the proxy's IP, or the server's IP.  Anyhow, guess that's a bit of a tangent.

Reply:
I struggle the most with this series of comments. As I understand it, if we have a proxy that takes care of resolving domains, we don't want to resolve them locally for the purposes of determining 1) what to create a connection with 2) WebSocket's "do we have a connection to this host" side table. These are the two use cases that "resolve an origin" is targeting and there's potentially dns-prefetch as a third use case in the future.

Reply:
It seems there are actually two concepts here to distinguish: 1) the set of IP addresses resolved from a hostname, and 2) the remote IP address of a network socket. This operation defines 1), but Private Network Access operates on 2).

Currently, even in the case of SOCKSv4 proxies where Chrome resolves hosts itself, Private Network Access checks apply to the proxy's IP address because that is the remote endpoint of the socket. If we are to specify the current behavior,  then I agree with @sleevi that this special case might be better handled at the connection establishment layer rather than at the resolution layer.

Now, I would be happy to improve the situation in Chrome by using the actually-resolved IP address in cases where it is available, because that would reduce the extent of proxy weirdness. I don't think I will tackle this in the near term - given the corner case it is, it seems low priority to me.

Reply:
Ah, that's the 4th use case, PNA. I haven't addressed that yet. For that we indeed need to add a field on connection and decide what value it should have. I suggest we try tackling that separately from this PR.

Reply:
I agree! If this language about proxies is useful for others, I'm happy to let it be :)